src/Disks/DiskLocal.cpp

```diff
@@ -317,64 +367,160 @@ SyncGuardPtr DiskLocal::getDirectorySyncGuard(const String & path) const
+    char errmsg[64];
+    if (::access(fpath, F_OK) == 0)
+    {
+        if (::remove(fpath) != 0)
         {
-            disk->reserved_bytes -= size;
+            LOG_WARNING(logger, "Fail to delete test file. path = {}, errno = {}, err = {}.", fpath, errno, strerror_r(errno, errmsg, 64));
+            return false;
         }
+    }
+    else
+    {
+        if (errno != ENOENT)
+        {
+            LOG_WARNING(logger, "Fail to access test file. path = {}, errno = {}, err = {}.", fpath, errno, strerror_r(errno, errmsg, 64));
+            return false;
+        }
+    }
+    int fd;
+    if ((fd = ::open(fpath, O_RDWR | O_CREAT | O_SYNC, S_IRUSR | S_IWUSR)) < 0)
+    {
+        LOG_WARNING(logger, "Fail to create test file. path={}.", fpath);
+        return false;
+    }
 
-        if (disk->reservation_count == 0)
-            LOG_ERROR(disk->log, "Unbalanced reservation count for disk '{}'.", disk->getName());
-        else
-            --disk->reservation_count;
+    bool read_write_ok = true;
+    if (disk_checker)
+    {
+        auto rw_test = [fd](char * buf, auto && func)
+        {
+            char * ptr = buf;
+            auto buf_size = DiskLocalCheckThread::BUF_SIZE;
+            size_t offset = 0;
+            while (buf_size > 0)
+            {
+                ssize_t rd_size = func(fd, ptr, buf_size, offset);
+                if (rd_size <= 0)
+                    return false;
+                buf_size -= rd_size;
+                ptr += rd_size;
+                offset += rd_size;
+            }
+            return true;
+        };
+
+        if (!rw_test(disk_checker->wbuf, ::pwrite))
+        {
+            LOG_WARNING(logger, "Fail to write test file. [file_name={}, err={}].", fpath, strerror_r(errno, errmsg, 64));
+            read_write_ok = false;
+        }
+        else if (!rw_test(disk_checker->rbuf, ::pread))
+        {
+            LOG_WARNING(logger, "Fail to read test file. [file_name={}, err={}].", fpath, strerror_r(errno, errmsg, 64));
+            read_write_ok = false;
+        }
+        else if (memcmp(disk_checker->wbuf, disk_checker->rbuf, DiskLocalCheckThread::BUF_SIZE) != 0)
+        {
+            LOG_WARNING(logger, "Inconsistent read and write over test file, [file_name = {}].", fpath);
+            read_write_ok = false;
+        }
     }
-    catch (...)
+    if (::close(fd) != 0)
     {
-        tryLogCurrentException(__PRETTY_FUNCTION__);
+        LOG_WARNING(logger, "Fail to close test file. [file_name={}].", fpath);
+        return false;
     }
+    if (::remove(fpath) != 0)
+    {
+        LOG_WARNING(logger, "Fail to delete test file. [err={}, path={}].", strerror_r(errno, errmsg, 64), fpath);
+        return false;
+    }
+    return read_write_ok;
+}
+
+bool DiskLocal::shouldRecover() const
+{
+    Poco::File disk{disk_path};
+    String file_path = disk_path + ".recover";
+    const char * fpath = file_path.c_str();
+    char errmsg[64];
+    if (::access(fpath, F_OK) == 0)
```

Comment:
Looks like worth adding some `SYSTEM` query for this?
And it seems that it does not do any "recovering", so maybe worth renaming to "rescan" or similar?

Reply:
Luckily we have a `system restart disk` after https://github.com/ClickHouse/ClickHouse/pull/23429

Reply:
Done with `system restart disk`